Sprint 2

Neagz opened the pull request:
Внедрение TypeScript.
Создание страницы со списком чатов и лентой переписки.
Добавлен компонентный подход в проект.
Реализация блока (Block) и Event Bus.
Сделан сбор данных из формы. В console.log выводиться объект со всеми заполненными полями формы.
Добавлена валидацию на все формы. Валидация работает по blur-событиям и второй раз проверяется при нажатии на submit.
Структурирован проект.
Настроены правильные экспорты и импорты.
Добавлен класс для работы с запросами.
Добавлен ESLint
Добавлен Stylelint

Neagz committed:
- Страница чатов
- eslint
- stylelint

Neagz committed:
- Страница чат
- eslint
- stylelint

Neagz committed:
- Страница чат
- eslint
- stylelint

Neagz committed:
Merge remote-tracking branch 'origin/sprint_2' into sprint_2

Neagz committed:
-Исправление страницы Чата
-Верстка с Hbs
-Code style

Neagz committed:
-Code style
-Dialog HBS
-Partial-block
-Паттерны

Neagz committed:
- Реализация блока (Block) и Event Bus;
- Тестовая страница с тестовой кнопкой событий

Neagz committed:
- Валидация
- Отправка Форм в Console.log

Neagz committed:
- svg icon_attach done

Neagz committed:
- Blur test

Neagz committed:
- Blur, submit
- Запросы
- ESLint

Neagz committed:
- newline

Neagz committed:
- Линтеры Тесты - done

Neagz committed:
- Линтеры Тесты - done

Neagz committed:
- Линтеры Тесты - done

Neagz committed:
- Линтеры Тесты - done

Neagz committed:
- Линтеры Тесты - done

Neagz committed:
- Линтеры Тесты - done

Neagz committed:
- Линтеры Тесты - done

Neagz committed:
- Линтеры Тесты - done

Neagz committed:
- Линтеры Тесты - done

Neagz committed:
- Линтеры Тесты - done

aromanovnik commented on src/core/block.ts:
```diff
@@ -0,0 +1,157 @@
+type BlockChildren = Record<string, Block | Block[]>;
+
+export class Block<P extends Record<string, unknown> = Record<string, unknown>> {
+    static EVENTS = {
```
**Можно лучше**: Добавление `as const` позволяет зафиксировать литеральные типы значений объекта и предотвратить их приведение к общему типу `string`. Это особенно полезно, если значения объекта должны использоваться как конкретные литералы, а не просто строки. Без использования `as const`, значения объекта воспринимаются TypeScript’ом как обычные строки, что может привести к потере строгости типизации. Подробнее: https://igorfonin.gitbook.io/typescript-book-ru/typescript-type-system/literals

<img width="484" alt="Снимок экрана 2025-07-06 в 23 11 01" src="https://github.com/user-attachments/assets/72175941-484f-4279-9cc4-a22a4249ac30" />
<img width="567" alt="Снимок экрана 2025-07-06 в 23 11 21" src="https://github.com/user-attachments/assets/be0aefba-dda1-4d30-92e5-ac6074ff4acc" />

aromanovnik commented on src/core/block.ts:
```diff
@@ -0,0 +1,157 @@
+
+    private _addEvents() {
+        const { events = {} } = this.props;
+        // Удаляем только существующие обработчики
```
**Можно лучше**: Удаление слушателей событий лучше вынести в отдельный метод, например `_removeEvents`, и вызывать его не только при ререндере, но и при размонтировании компонента. Это поможет избежать утечек памяти и неожиданных побочных эффектов, особенно если компонент может монтироваться и размонтироваться несколько раз.

aromanovnik commented on src/core/httpTransport.ts:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    get(url: string, data?: RequestData): Promise<XMLHttpRequest> {
+        return this.request('GET', url, data);
```
**Надо исправить**: Согласно заданию, HTTP-методы (`GET`, `POST`, `PUT`, `DELETE`) должны быть вынесены в `enum` или `const`, чтобы избежать “магических строк” в коде и упростить их переиспользование и поддержку. Сейчас методы указаны напрямую, что противоречит требованиям.

aromanovnik commented on src/core/httpTransport.ts:
```diff
@@ -0,0 +1,75 @@
+
+            xhr.onload = () => {
+                // Успешные статусы 200-299
+                if (xhr.status >= 200 && xhr.status < 300) {
```
**Можно лучше**: Для удобства и избежания ошибок в коде статусы HTTP-ответов лучше вынести в `enum`. Это упрощает поддержку и позволяет избежать случайных опечаток при использовании числовых значений. Например: `enum HttpStatus { OK = 200, …}`

aromanovnik commented on src/core/block.ts:
```diff
@@ -0,0 +1,157 @@
+// Тип для дочерних компонентов
+type BlockChildren = Record<string, Block | Block[]>;
+
+export class Block<P extends Record<string, unknown> = Record<string, unknown>> {
```
**Отлично**: Свойство `props` строго типизировано через дженерик.

aromanovnik replied:
**Можно лучше:** Чтобы предотвратить создание экземпляров класса напрямую, необходимо использовать ключевое слово `abstract` перед определением класса. Это сделает класс абстрактным и позволит использовать его только как базовый класс для наследования. <https://www.typescriptlang.org/docs/handbook/2/classes.html#abstract-classes-and-members>

aromanovnik commented on src/core/httpTransport.ts:
```diff
@@ -0,0 +1,75 @@
+        return this.request('PUT', url, data);
+    }
+
+    delete(url: string, data?: RequestData): Promise<XMLHttpRequest> {
```
**Можно лучше**: Если типизация повторяется в каждом методе (функции), то лучше типизировать сам метод (функцию), чтобы избежать дублирования типизации аргументов. Или можно использовать фабричный метод, который будет возвращать соответствующий функционал в зависимости от типа запроса. Это улучшит структуру кода, повысит его повторное использование и упростит изменения в будущем, поскольку все изменения будут централизованы в одном месте.

```typescript
type HTTPMethod = <R = unknown>(url: string, options?: Partial<RequestOptions<QueryParams>>) => Promise<R>;


export class HTTPTransport {
  // Фабричный метод для создания HTTP-методов
  private createMethod(method: METHODS): HTTPMethod {
    return (url, options = {}) => this.request(url, { ...options, method });
  }

  // Методы HTTP
  protected readonly get = this.createMethod(METHODS.GET);

  protected readonly put = this.createMethod(METHODS.PUT);

  protected readonly post = this.createMethod(METHODS.POST);

  protected readonly delete = this.createMethod(METHODS.DELETE);
  
  // Реализация метода request
  private request<R>(
    url: string,
    options: RequestOptions<QueryParams>,
  ): Promise<R> {
    // Логика выполнения HTTP-запроса
  }
}
```

aromanovnik commented on package.json:
```diff
@@ -6,16 +6,41 @@
   "scripts": {
     "start": "vite",
```
**Надо исправить**: Команда `vite` запускает сервер в **development (dev) режиме**, где доступна горячая перезагрузка, подробные сообщения об ошибках и исходные карты, но производительность и поведение могут отличаться от продакшена. Для сборки проекта следует использовать `vite build`, а для запуска локального сервера **не** в DEV-режиме — `vite preview`. Скрипты `start` и `dev` могут выглядеть так:

```javascript
"start": "vite build && vite preview",
"dev": "vite"
```

Не забудьте настроить `preview` на порт `3000`. <https://vite.dev/config/preview-options>

aromanovnik commented on src/main.ts:
```diff
@@ -1,55 +1,110 @@
-document.addEventListener('DOMContentLoaded', init);
+// Глобальная навигация по кликам
+document.addEventListener('click', (e) => {
+  const target = e.target as HTMLElement;
```
**Можно лучше:** Переменная `target` имеет тип `EventTarget | null`, но с помощью `as` утверждается, что это `HTMLElement`. Это может привести к ошибкам, если `target` окажется не `HTMLElement`. Лучше уточнить типы и добавить условия проверки, например, через проверку `target instanceof HTMLElement`. Это улучшит надежность кода, так как условие явно проверяет, что объект действительно является `HTMLElement`, прежде чем выполнять с ним операции.

aromanovnik commented on src/pages/profile_edit/profile_edit.ts:
```diff
@@ -0,0 +1,279 @@
+    [key: string]: unknown;
+}
+
+export class ProfileEditPage extends Block<ProfileEditPageProps> {
```
**Надо исправить:** По требованиям задания 1 спринта, должен быть инпут для выбора аватарки, с атрибутом `name`, равным `avatar`.

aromanovnik commented on src/pages/list/list.hbs:
```diff
@@ -1,17 +1,65 @@
+            </div>
+        </div>
+        <div class="chat__main">
+            <div class="chat-header">
```
**Можно лучше**: Использование семантической разметки, такой как `<header>`, `<article>`, `<section>`, `<footer>`, помогает сделать сайт более доступным для пользователей с ограниченными возможностями и улучшить его ранжирование в поисковых системах. Это обеспечивает лучшую структуру и понятность содержимого для браузеров и поисковых роботов. Подробнее: <https://htmlacademy.ru/blog/html/semantics>

aromanovnik reviewed (changes requested):
Здравствуйте, Андрей! 👋 *(Нужно развернуть общий комментарий ↓)*

**Работа выполнена очень хорошо**:

✅ Линтеры настроены и выполняются без ошибок.
✅ Проект полностью типизирован, `any` отсутствуют.
✅ Хорошо написаны регулярные выражения, учтена буква `ё`.
✅ `Block.props` строго типизированы.
✅ Обработчики отправки формы грамотно привязаны к событию `submit`.

Код чистый, структурированный!
Отличная работа!

Тем не менее, есть несколько замечаний, которые нужно доработать. 🔧 Более детальная информация указана в комментариях к коду.

**Что нужно исправить**:

* Согласно заданию, HTTP-методы (`GET`, `POST`, `PUT`, `DELETE`) должны быть вынесены в `enum` или `const`, чтобы избежать “магических строк” в коде.
* Команда `vite` запускает сервер в **development (dev) режиме**, где доступна горячая перезагрузка, подробные сообщения об ошибках и исходные карты, но производительность и поведение могут отличаться от продакшена. 
* По требованиям задания 1 спринта, должен быть инпут для выбора аватарки, с атрибутом `name`, равным `avatar`.

**Что можно улучшить**:

* Добавление `as const` позволяет зафиксировать литеральные типы значений объекта и предотвратить их приведение к общему типу `string`. 
* Удаление слушателей событий лучше вынести в отдельный метод, например `_removeEvents`, и вызывать его не только при ререндере, но и при размонтировании компонента.
* Для удобства и избежания ошибок в коде статусы HTTP-ответов лучше вынести в `enum`.
* Чтобы предотвратить создание экземпляров класса напрямую, необходимо использовать ключевое слово `abstract` перед определением класса. 
* Если типизация повторяется в каждом методе (функции), то лучше типизировать сам метод (функцию), чтобы избежать дублирования типизации аргументов.
* Переменная `target` имеет тип `EventTarget | null`, но с помощью `as` утверждается, что это `HTMLElement`. Это может привести к ошибкам, если `target` окажется не `HTMLElement`.
* Использование семантической разметки, такой как `<header>`, `<article>`, `<section>`, `<footer>`, помогает сделать сайт более доступным для пользователей с ограниченными возможностями и улучшить его ранжирование в поисковых системах.

Напоминаю, что работа может быть принята только после исправления всех критических замечаний **Надо исправить**.

Удачного рефакторинга кода.

Neagz committed:
- Типизация
- Вынос методов и статусов в enum
- Коррекция запуска сборки
- Новый компонент avatar-input
- Семантическая разметка

aromanovnik reviewed (approved):
🎉 **Поздравляю! Работа принята!**

Все недочёты успешно исправлены, и результат получился просто отличным!

Вы проделали огромную работу, и это действительно впечатляет!
Желаю успехов в дальнейшей учёбе и новых достижений! 🚀
